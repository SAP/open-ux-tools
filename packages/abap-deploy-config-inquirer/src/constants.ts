export const ABAP_PACKAGE_SEARCH_MAX_RESULTS = 50;
export const DEFAULT_PACKAGE_ABAP = '$tmp';
