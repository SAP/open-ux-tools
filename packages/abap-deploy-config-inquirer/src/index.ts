import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
