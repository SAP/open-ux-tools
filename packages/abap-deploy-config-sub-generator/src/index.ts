export { getAbapQuestions } from './app/questions';
export { indexHtmlExists, showOverwriteQuestion } from './utils';
export { AbapDeployConfigOptions, DeployProjectType } from './app/types';
