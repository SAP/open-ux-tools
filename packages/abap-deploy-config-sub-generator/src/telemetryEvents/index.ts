/**
 * Event names for telemetry for the abap deploy config generator
 */
export enum EventName {
    DEPLOY_CONFIG = 'DEPLOY_CONFIG'
}
