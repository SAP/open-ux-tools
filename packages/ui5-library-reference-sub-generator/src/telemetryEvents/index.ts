/**
 * Event names for telemetry for the reference library generator
 */
export enum EventName {
    LIB_REFERENCE_ADDED = 'LIB_REFERENCE_ADDED'
}
