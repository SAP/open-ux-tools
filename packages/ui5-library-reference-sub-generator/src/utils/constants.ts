export const generatorTitle = 'SAP Fiori Reusable Library';
export const prompts = [
    {
        name: 'Create a Reference',
        description: 'Add library reference to an SAP Fiori application'
    }
];
