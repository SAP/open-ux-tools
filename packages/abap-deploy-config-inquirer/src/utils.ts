import { isAppStudio, listDestinations } from '@sap-ux/btp-utils';
import { getService } from '@sap-ux/store';
import { PackageInputChoices, TargetSystemType, TransportChoices } from './types';
import { getTransportConfigInstance } from './service-provider-utils';
import { t } from './i18n';
import LoggerHelper from './logger-helper';
import { listPackages } from './validator-utils';
import type {
    AbapDeployConfigAnswers,
    AbapDeployConfigAnswersInternal,
    BackendTarget,
    Credentials,
    InitTransportConfigResult,
    SystemConfig
} from './types';
import type { BackendSystem, BackendSystemKey } from '@sap-ux/store';
import type { Destinations, Destination } from '@sap-ux/btp-utils';
import { CREATE_TR_DURING_DEPLOY } from './constants';

let cachedDestinations: Destinations = {};
let cachedBackendSystems: BackendSystem[] = [];

/**
 * Retrieve the ABAP systems from the store or the destinations from BTP.
 *
 * @returns destinations or backend systems from the store
 */
export async function getAbapSystems(): Promise<{
    destinations: Destinations | undefined;
    backendSystems: BackendSystem[] | undefined;
}> {
    let destinations;
    let backendSystems;

    if (isAppStudio()) {
        destinations = await listDestinations({
            stripS4HCApiHosts: true
        });
        cachedDestinations = destinations;
    } else {
        const systemStore = await getService<BackendSystem, BackendSystemKey>({
            logger: LoggerHelper.logger,
            entityName: 'system'
        });
        backendSystems = await systemStore.getAll();
        cachedBackendSystems = backendSystems;
    }

    return { destinations, backendSystems };
}

/**
 * Retrieve a specific Destination, based on the destination name.
 *
 * @param destination - destination name
 * @returns destination if found
 */
export function findDestination(destination: string): Destination | undefined {
    return cachedDestinations?.[destination];
}

/**
 * Retrieve a specific BackendSystem, based on the URL.
 *
 * @param backendUrl - backend system URL
 * @returns backend system if found
 */
export function findBackendSystemByUrl(backendUrl: string): BackendSystem | undefined {
    return cachedBackendSystems?.find((backend: BackendSystem) => backend.url === backendUrl);
}

/**
 * Check if the current system is the same as the one in the answers.
 *
 * @param abapSystem - system configuration
 * @param url - url
 * @param client - client
 * @param destination - destination
 * @returns true if the system is the same
 */
export function isSameSystem(abapSystem?: SystemConfig, url?: string, client?: string, destination?: string): boolean {
    return Boolean(
        (abapSystem?.url &&
            abapSystem.url.trim()?.replace(/\/$/, '') === url?.trim()?.replace(/\/$/, '') &&
            abapSystem.client === client) ??
            (!!abapSystem?.destination && destination === abapSystem?.destination)
    );
}

/**
 * Get transport configuration from the backend.
 *
 * @param transportConfigParams - transport configuration parameters
 * @param transportConfigParams.backendTarget - backend target from prompt options
 * @param transportConfigParams.scp - scp
 * @param transportConfigParams.url - url
 * @param transportConfigParams.client - client
 * @param transportConfigParams.destination - destination
 * @param transportConfigParams.errorHandler - error handler
 * @param transportConfigParams.credentials - user credentials
 * @returns transport configuration
 */
export async function initTransportConfig({
    backendTarget,
    scp,
    url,
    client,
    destination,
    credentials,
    errorHandler
}: {
    backendTarget?: BackendTarget;
    scp?: boolean;
    url?: string;
    client?: string;
    destination?: string;
    credentials?: Credentials;
    errorHandler: (errorMessage: string) => void;
}): Promise<InitTransportConfigResult> {
    let result: InitTransportConfigResult = {};
    if (!url && !destination) {
        return result;
    }

    const systemConfig = {
        url,
        client,
        destination
    };

    try {
        result = await getTransportConfigInstance({
            backendTarget,
            scp,
            credentials,
            systemConfig
        });
    } catch (e) {
        result.error = e;
    }

    if (result.error) {
        errorHandler(result.error);
        LoggerHelper.logger.debug(
            t('errors.debugAbapTargetSystem', { method: 'initTransportConfig', error: result.error })
        );
    }

    return result;
}

/**
 * Querying package names that match the user input.
 *
 * @param input - user input
 * @param inputSystemConfig System configuration extracted from user answers for establishing backend connection
 * @param backendTarget - backend target from abap deploy config prompt options
 * @returns list of package names
 */
export async function queryPackages(
    input: string,
    inputSystemConfig: SystemConfig,
    backendTarget?: BackendTarget
): Promise<string[]> {
    const uppercaseInput = (input ?? '').toUpperCase();
    return listPackages(uppercaseInput, inputSystemConfig, backendTarget);
}

/**
 * Determines the package from the various package related prompts.
 *
 * @param previousAnswers - previous answers
 * @param statePackage - package from state
 * @returns package name
 */
export function getPackageAnswer(previousAnswers?: AbapDeployConfigAnswersInternal, statePackage?: string): string {
    // Older versions of YUI do not have a packageInputChoice question
    return statePackage || previousAnswers?.packageInputChoice === PackageInputChoices.ListExistingChoice
        ? previousAnswers?.packageAutocomplete ?? ''
        : previousAnswers?.packageManual ?? '';
}

/**
 * Determines the transport request from the various transport related prompts.
 *
 * @param promptAnswers - previous answers
 * @returns transport request
 */
export function getTransportAnswer(promptAnswers?: AbapDeployConfigAnswersInternal): string {
    return (
        promptAnswers?.transportManual ||
        promptAnswers?.transportFromList ||
        promptAnswers?.transportCreated ||
        (promptAnswers?.transportInputChoice === TransportChoices.CreateDuringDeployChoice
            ? CREATE_TR_DURING_DEPLOY
            : '')
    );
}

/**
 * Check if the transport matches placeholder used to create transport request number during actual deploy process.
 *
 * @param transport - existing transport
 * @returns true if transport setting is set to 'CreateDuringDeployChoice'.
 */
export function useCreateTrDuringDeploy(transport?: string): boolean {
    return transport === CREATE_TR_DURING_DEPLOY;
}

/**
 * Determines the url from the various sources.
 *
 * @param answers - internal abap deploy config answers
 * @param stateUrl - url from state
 * @returns url if found
 */
function getUrlAnswer(answers: AbapDeployConfigAnswersInternal, stateUrl?: string): string {
    let url = answers.url;
    if (stateUrl) {
        url = stateUrl;
    }
    return url;
}

/**
 * Convert internal answers to external answers to be used for writing deploy config.
 *
 * @param answers - internal abap deploy config answers, direct from prompting
 * @param state - partial internal abap deploy config answers derived from the state
 * @returns - external abap deploy config answers
 */
export function reconcileAnswers(
    answers: AbapDeployConfigAnswersInternal,
    state: Partial<AbapDeployConfigAnswersInternal>
): AbapDeployConfigAnswers {
    // Add dervied service answers to the answers object
    answers = Object.assign(answers, state);

    const reconciledAnswers: AbapDeployConfigAnswers = {
        url: getUrlAnswer(answers, state.url),
        package: getPackageAnswer(answers, state.package)
    };

    if (answers.destination) {
        reconciledAnswers.destination = answers.destination;
    }

    if (answers.targetSystem && answers.targetSystem !== TargetSystemType.Url) {
        reconciledAnswers.url = answers.targetSystem;
    }

    if (answers.client || state.client) {
        reconciledAnswers.client = answers.client || state.client;
    }

    if (answers.scp || state.scp) {
        reconciledAnswers.scp = true;
    }

    if (answers.ui5AbapRepo) {
        reconciledAnswers.ui5AbapRepo = answers.ui5AbapRepo;
    }

    if (answers.description) {
        reconciledAnswers.description = answers.description;
    }

    const transportAnswer = getTransportAnswer(answers);
    if (transportAnswer) {
        reconciledAnswers.transport = transportAnswer;
    }

    if (answers.index !== undefined) {
        reconciledAnswers.index = answers.index;
    }

    if (answers.overwrite !== undefined) {
        reconciledAnswers.overwrite = answers.overwrite;
    }

    return reconciledAnswers;
}
