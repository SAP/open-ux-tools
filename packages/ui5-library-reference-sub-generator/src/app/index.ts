import Generator from 'yeoman-generator';
import ReferenceLibGenLogger from '../utils/logger';
import { generatorTitle, prompts } from '../utils/constants';
import { workspace } from 'vscode';
import { Prompts, AppWizard, MessageType } from '@sap-devx/yeoman-ui-types';
import { generate } from '@sap-ux/ui5-library-reference-writer';
import {
    isExtensionInstalled,
    sendTelemetry,
    TelemetryHelper,
    YUI_EXTENSION_ID,
    YUI_MIN_VER_FILES_GENERATED_MSG
} from '@sap-ux/fiori-generator-shared';
import { isInternalFeaturesSettingEnabled } from '@sap-ux/feature-toggle';
import { prompt } from '@sap-ux/ui5-library-reference-inquirer';
import { t } from '../utils/i18n';
import { EventName } from '../telemetryEvents';
import type { UI5ReferenceLibGenerator } from './types';
import type { ReuseLibConfig } from '@sap-ux/ui5-library-reference-writer';
import type { YeomanEnvironment } from '@sap-ux/fiori-generator-shared';
import type { InquirerAdapter, UI5LibraryReferenceAnswers } from '@sap-ux/ui5-library-reference-inquirer';

/**
 * Generator for adding a UI5 library reference to a Fiori application.
 *
 * @extends Generator
 */
export default class extends Generator implements UI5ReferenceLibGenerator {
    answers: UI5LibraryReferenceAnswers = {};
    prompts: Prompts;
    appWizard: AppWizard;
    basePath: string;
    vscode: unknown;
    setPromptsCallback: (fn: any) => void;

    /**
     * Constructor for the generator.
     *
     * @param args - arguments passed to the generator
     * @param opts - options passed to the generator
     */
    constructor(args: string | string[], opts: Generator.GeneratorOptions) {
        super(args, opts);

        this.appWizard = opts.appWizard || AppWizard.create(opts);
        this.vscode = opts.vscode;
        ReferenceLibGenLogger.configureLogging(
            this.options.logger,
            this.rootGeneratorName(),
            this.log,
            this.options.vscode,
            this.options.logLevel
        );

        this.appWizard.setHeaderTitle(generatorTitle);
        this.prompts = new Prompts(prompts);
        this.setPromptsCallback = (fn): void => {
            if (this.prompts) {
                this.prompts.setCallback(fn);
            }
        };
    }

    public async initializing(): Promise<void> {
        await TelemetryHelper.initTelemetrySettings({
            consumerModule: {
                name: '@sap-ux/ui5-library-reference-sub-generator',
                version: this.rootGeneratorVersion()
            },
            internalFeature: isInternalFeaturesSettingEnabled(),
            watchTelemetrySettingStore: false
        });
        if ((this.env as unknown as YeomanEnvironment).conflicter) {
            (this.env as unknown as YeomanEnvironment).conflicter.force = true;
        }
    }

    public async prompting(): Promise<void> {
        let inquirerAdaptor;

        if ((this.env as unknown as YeomanEnvironment)?.adapter?.actualAdapter) {
            inquirerAdaptor = (this.env as unknown as YeomanEnvironment).adapter.actualAdapter;
        } else {
            inquirerAdaptor = this.env?.adapter;
        }
        const answers = await prompt(workspace.workspaceFolders, inquirerAdaptor as InquirerAdapter);
        Object.assign(this.answers, answers);
    }

    public async writing(): Promise<void> {
        const reuseLibConfigs: ReuseLibConfig[] = [];
        if (this.answers.referenceLibraries) {
            for (const lib of this.answers.referenceLibraries) {
                reuseLibConfigs.push({
                    name: lib.name,
                    path: lib.path,
                    type: lib.type,
                    uri: lib.uri
                });
            }
        }

        try {
            if (this.answers.targetProjectFolder) {
                await generate(this.answers.targetProjectFolder, reuseLibConfigs, this.fs);
            }
        } catch (e) {
            ReferenceLibGenLogger.logger.error(e);
            throw new Error(t('error.updatingApp'));
        }
    }

    end(): void {
        try {
            if (isExtensionInstalled(this.vscode, YUI_EXTENSION_ID, YUI_MIN_VER_FILES_GENERATED_MSG)) {
                this.appWizard?.showInformation(t('info.filesGenerated'), MessageType.notification);
            }
            const telemetryData = TelemetryHelper.createTelemetryData();
            if (telemetryData) {
                sendTelemetry(EventName.LIB_REFERENCE_ADDED, telemetryData, this.answers.targetProjectFolder).catch(
                    (error) => {
                        ReferenceLibGenLogger.logger.error(t('error.telemetry', { error }));
                    }
                );
            }
        } catch (error) {
            ReferenceLibGenLogger.logger.error(t('error.endPhase'));
        }
    }
}
