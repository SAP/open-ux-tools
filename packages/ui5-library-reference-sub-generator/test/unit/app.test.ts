import yeomanTest from 'yeoman-test';
import path from 'path';
import fs from 'fs';
import fsextra from 'fs-extra';
import RefLibGenerator from '../../src/app';
import { reuseLibs } from './util/constants';
import { rimraf } from 'rimraf';
import { EventName } from '../../src/telemetryEvents';
import { t } from '../../src/utils/i18n';
import { MessageType } from '@sap-devx/yeoman-ui-types';
import * as fioriGenShared from '@sap-ux/fiori-generator-shared';
import * as ui5LibRefWriter from '@sap-ux/ui5-library-reference-writer';
import * as ui5LibraryRefInquirer from '@sap-ux/ui5-library-reference-inquirer';
import type { UI5LibraryReferenceAnswers } from '@sap-ux/ui5-library-reference-inquirer';

const testOutputDir = path.join(__dirname, '..', 'test-output', 'app');
const originalCwd = process.cwd();
const refLibGenPath = path.join(__dirname, '../../src/app');
let yoEnv4 = false;

jest.mock('@sap-ux/fiori-generator-shared', () => ({
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    ...(jest.requireActual('@sap-ux/fiori-generator-shared') as {}),
    TelemetryHelper: {
        initTelemetrySettings: jest.fn(),
        createTelemetryData: jest.fn().mockReturnValue({
            OperatingSystem: 'CLI',
            Platform: 'darwin'
        })
    },
    sendTelemetry: jest.fn(),
    isExtensionInstalled: jest.fn().mockReturnValue(true)
}));

jest.mock('yeoman-test/lib/adapter', () => ({
    TestAdapter: function TestAdapter() {
        const testAdapter = new (jest.requireActual('yeoman-test/lib/adapter').TestAdapter)();
        if (yoEnv4) {
            Object.assign(testAdapter, { actualAdapter: {} });
        }
        return testAdapter;
    },
    DummyPrompt: jest.requireActual('yeoman-test/lib/adapter').DummyPrompt
}));

afterAll(() => {
    process.chdir(originalCwd); // Generation changes the cwd, this breaks sonar report so we restore later
    rimraf.sync(testOutputDir);
});

describe('Test reference generator', () => {
    jest.setTimeout(60000);
    it('should run the generator', async () => {
        const sendTelemetrySpy = jest.spyOn(fioriGenShared, 'sendTelemetry');
        const testProjectPath = path.join(testOutputDir, 'test_project');
        fs.mkdirSync(testOutputDir, { recursive: true });
        fs.mkdirSync(testProjectPath);
        fsextra.copySync(path.join(__dirname, 'sample/test_project_lrop_v2'), testProjectPath);
        const showInformationSpy = jest.fn();
        const mockAppWizard = {
            setHeaderTitle: jest.fn(),
            showInformation: showInformationSpy
        };

        await yeomanTest
            .run(RefLibGenerator, {
                resolved: refLibGenPath
            })
            .withOptions({
                appWizard: mockAppWizard
            })
            .withPrompts({
                targetProjectFolder: testProjectPath,
                source: 'Workspace',
                referenceLibraries: reuseLibs.map((lib) => lib.value)
            });

        const manifest = await fs.promises.readFile(path.join(testProjectPath, 'webapp/manifest.json'), {
            encoding: 'utf8'
        });
        const updatedManifest = JSON.parse(manifest);
        expect(updatedManifest).toMatchSnapshot();
        expect(sendTelemetrySpy).toBeCalledWith(
            EventName.LIB_REFERENCE_ADDED,
            expect.objectContaining({
                OperatingSystem: 'CLI',
                Platform: 'darwin'
            }),
            testProjectPath
        );
        expect(showInformationSpy).toHaveBeenCalledWith(t('info.filesGenerated'), MessageType.notification);
    });
    it('should run the generator custom webapp path', async () => {
        const testProject = path.join(testOutputDir, 'test_project_lrop_v2_custom_webapp_path');
        fs.mkdirSync(testOutputDir, { recursive: true });
        fs.mkdirSync(testProject);
        fsextra.copySync(path.join(__dirname, 'sample/test_project_lrop_v2_custom_webapp_path'), testProject);

        await yeomanTest
            .run(RefLibGenerator, {
                resolved: refLibGenPath
            })
            .withPrompts({
                targetProjectFolder: testProject,
                source: 'Workspace',
                referenceLibraries: reuseLibs.map((lib) => lib.value)
            });

        const manifest = await fs.promises.readFile(path.join(testProject, 'src/main/webapp/manifest.json'), {
            encoding: 'utf8'
        });
        const updatedManifest = JSON.parse(manifest);
        expect(updatedManifest).toMatchSnapshot();
    });

    it('should throw error when writing fails', async () => {
        const testProject = path.join(testOutputDir, 'test_project_lrop_v2_custom_webapp_path_2');
        fs.mkdirSync(testOutputDir, { recursive: true });
        fs.mkdirSync(testProject);
        fsextra.copySync(path.join(__dirname, 'sample/test_project_lrop_v2_custom_webapp_path'), testProject);
        jest.spyOn(ui5LibRefWriter, 'generate').mockRejectedValueOnce(new Error('Error'));

        try {
            await yeomanTest
                .run(RefLibGenerator, {
                    resolved: refLibGenPath
                })
                .withPrompts({
                    targetProjectFolder: testProject,
                    source: 'Workspace',
                    referenceLibraries: reuseLibs.map((lib) => lib.value)
                });
        } catch (e) {
            expect(e.message).toBe(t('error.updatingApp'));
        }
    });

    it('prompting with yeoman-environment@^4 default adaptor (yo@5 support)', async () => {
        const testProject = path.join(testOutputDir, 'test_project_lrop_v2_custom_webapp_path_2');
        const promptAnswers = {
            targetProjectFolder: testProject,
            source: 'Workspace',
            referenceLibraries: reuseLibs.map((lib) => lib.value)
        } as unknown as ui5LibraryRefInquirer.UI5LibraryReferenceAnswers;
        jest.spyOn(ui5LibRefWriter, 'generate').mockResolvedValueOnce({} as any);
        const promptSpy = jest
            .spyOn(ui5LibraryRefInquirer, 'prompt')
            .mockResolvedValue(promptAnswers as unknown as UI5LibraryReferenceAnswers);
        // Use the mocked adapter representing yeoman-environment@4
        yoEnv4 = true;
        await expect(
            yeomanTest
                .run(RefLibGenerator, {
                    resolved: refLibGenPath
                })
                .withOptions({ skipInstall: false })
        ).resolves.not.toThrow();
        expect(promptSpy).toHaveBeenCalled();
    });
});
