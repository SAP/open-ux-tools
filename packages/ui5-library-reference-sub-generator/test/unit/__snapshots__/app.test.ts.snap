// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test reference generator should run the generator 1`] = `
{
  "_version": "1.17.0",
  "sap.app": {
    "applicationVersion": {
      "version": "1.0.0",
    },
    "dataSources": {
      "SEPMRA_PROD_MAN_ANNO_MDL": {
        "settings": {
          "localUri": "localService/SEPMRA_PROD_MAN_ANNO_MDL.xml",
        },
        "type": "ODataAnnotation",
        "uri": "/sap/opu/odata/IWFND/CATALOGSERVICE;v=2/Annotations(TechnicalName='SEPMRA_PROD_MAN_ANNO_MDL',Version='0001')/$value/",
      },
      "annotation": {
        "settings": {
          "localUri": "annotations/annotation.xml",
        },
        "type": "ODataAnnotation",
        "uri": "annotations/annotation.xml",
      },
      "mainService": {
        "settings": {
          "annotations": [
            "SEPMRA_PROD_MAN_ANNO_MDL",
            "annotation",
          ],
          "localUri": "localService/metadata.xml",
        },
        "type": "OData",
        "uri": "/sap/opu/odata/sap/SEPMRA_PROD_MAN/",
      },
    },
    "description": "{{appDescription}}",
    "i18n": "i18n/i18n.properties",
    "id": "testNameSpace.testprojectlropv2",
    "offline": false,
    "resources": "resources.json",
    "sourceTemplate": {
      "id": "ui5template.smartTemplate",
      "toolsId": "4e5poddc-a666-4465-89eb-092e768dfecc",
      "version": "1.40.12",
    },
    "tags": {
      "keywords": [],
    },
    "title": "{{appTitle}}",
    "type": "application",
  },
  "sap.fiori": {
    "archeType": "transactional",
    "registrationIds": [],
  },
  "sap.platform.abap": {
    "uri": "",
  },
  "sap.platform.hcp": {
    "uri": "",
  },
  "sap.ui": {
    "deviceTypes": {
      "desktop": true,
      "phone": true,
      "tablet": true,
    },
    "icons": {
      "favIcon": "",
      "icon": "",
      "phone": "",
      "phone@2": "",
      "tablet": "",
      "tablet@2": "",
    },
    "supportedThemes": [
      "sap_hcb",
      "sap_belize",
      "sap_fiori_3",
    ],
    "technology": "UI5",
  },
  "sap.ui.generic.app": {
    "_version": "1.3.0",
    "pages": {
      "ListReport|SEPMRA_C_PD_Product": {
        "component": {
          "list": true,
          "name": "sap.suite.ui.generic.template.ListReport",
          "settings": {
            "condensedTableLayout": true,
            "enableTableFilterInPageVariant": true,
            "filterSettings": {
              "dateSettings": {
                "useDateRange": true,
              },
            },
            "smartVariantManagement": true,
          },
        },
        "entitySet": "SEPMRA_C_PD_Product",
        "pages": {
          "ObjectPage|SEPMRA_C_PD_Product": {
            "component": {
              "name": "sap.suite.ui.generic.template.ObjectPage",
            },
            "defaultLayoutTypeIfExternalNavigation": "MidColumnFullScreen",
            "entitySet": "SEPMRA_C_PD_Product",
            "pages": {
              "ObjectPage|to_ProductSalesData": {
                "component": {
                  "name": "sap.suite.ui.generic.template.ObjectPage",
                },
                "defaultLayoutTypeIfExternalNavigation": "MidColumnFullScreen",
                "entitySet": "SEPMRA_C_PD_ProductSalesData",
                "navigationProperty": "to_ProductSalesData",
              },
            },
          },
        },
      },
    },
    "settings": {
      "considerAnalyticalParameters": true,
      "forceGlobalRefresh": false,
      "objectPageHeaderType": "Dynamic",
      "showDraftToggle": false,
    },
  },
  "sap.ui5": {
    "contentDensities": {
      "compact": true,
      "cozy": true,
    },
    "dependencies": {
      "components": {},
      "libs": {
        "sap.suite.ui.generic.template": {
          "lazy": false,
        },
        "sap.ui.core": {
          "lazy": false,
        },
        "sap.ui.generic.app": {
          "lazy": false,
        },
        "se.mi.plm.attachmentservice": {
          "lazy": false,
        },
      },
      "minUI5Version": "1.65.0",
    },
    "extends": {
      "extensions": {},
    },
    "models": {
      "": {
        "dataSource": "mainService",
        "preload": true,
        "settings": {
          "defaultBindingMode": "TwoWay",
          "defaultCountMode": "Inline",
          "metadataUrlParams": {
            "sap-value-list": "none",
          },
          "refreshAfterChange": false,
        },
      },
      "@i18n": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/i18n.properties",
      },
      "i18n": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/i18n.properties",
      },
      "i18n|sap.suite.ui.generic.template.ListReport|SEPMRA_C_PD_Product": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/ListReport/SEPMRA_C_PD_Product/i18n.properties",
      },
      "i18n|sap.suite.ui.generic.template.ObjectPage|SEPMRA_C_PD_Product": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/ObjectPage/SEPMRA_C_PD_Product/i18n.properties",
      },
      "i18n|sap.suite.ui.generic.template.ObjectPage|SEPMRA_C_PD_ProductSalesData": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/ObjectPage/SEPMRA_C_PD_ProductSalesData/i18n.properties",
      },
    },
    "resources": {
      "css": [],
      "js": [],
    },
  },
}
`;

exports[`Test reference generator should run the generator custom webapp path 1`] = `
{
  "_version": "1.17.0",
  "sap.app": {
    "applicationVersion": {
      "version": "1.0.0",
    },
    "dataSources": {
      "SEPMRA_PROD_MAN_ANNO_MDL": {
        "settings": {
          "localUri": "localService/SEPMRA_PROD_MAN_ANNO_MDL.xml",
        },
        "type": "ODataAnnotation",
        "uri": "/sap/opu/odata/IWFND/CATALOGSERVICE;v=2/Annotations(TechnicalName='SEPMRA_PROD_MAN_ANNO_MDL',Version='0001')/$value/",
      },
      "annotation": {
        "settings": {
          "localUri": "annotations/annotation.xml",
        },
        "type": "ODataAnnotation",
        "uri": "annotations/annotation.xml",
      },
      "mainService": {
        "settings": {
          "annotations": [
            "SEPMRA_PROD_MAN_ANNO_MDL",
            "annotation",
          ],
          "localUri": "localService/metadata.xml",
        },
        "type": "OData",
        "uri": "/sap/opu/odata/sap/SEPMRA_PROD_MAN/",
      },
    },
    "description": "{{appDescription}}",
    "i18n": "i18n/i18n.properties",
    "id": "testNameSpace.testprojectlropv2",
    "offline": false,
    "resources": "resources.json",
    "sourceTemplate": {
      "id": "ui5template.smartTemplate",
      "version": "1.40.12",
    },
    "tags": {
      "keywords": [],
    },
    "title": "{{appTitle}}",
    "type": "application",
  },
  "sap.fiori": {
    "archeType": "transactional",
    "registrationIds": [],
  },
  "sap.platform.abap": {
    "uri": "",
  },
  "sap.platform.hcp": {
    "uri": "",
  },
  "sap.ui": {
    "deviceTypes": {
      "desktop": true,
      "phone": true,
      "tablet": true,
    },
    "icons": {
      "favIcon": "",
      "icon": "",
      "phone": "",
      "phone@2": "",
      "tablet": "",
      "tablet@2": "",
    },
    "supportedThemes": [
      "sap_hcb",
      "sap_belize",
      "sap_fiori_3",
    ],
    "technology": "UI5",
  },
  "sap.ui.generic.app": {
    "_version": "1.3.0",
    "pages": {
      "ListReport|SEPMRA_C_PD_Product": {
        "component": {
          "list": true,
          "name": "sap.suite.ui.generic.template.ListReport",
          "settings": {
            "condensedTableLayout": true,
            "enableTableFilterInPageVariant": true,
            "filterSettings": {
              "dateSettings": {
                "useDateRange": true,
              },
            },
            "smartVariantManagement": true,
          },
        },
        "entitySet": "SEPMRA_C_PD_Product",
        "pages": {
          "ObjectPage|SEPMRA_C_PD_Product": {
            "component": {
              "name": "sap.suite.ui.generic.template.ObjectPage",
            },
            "defaultLayoutTypeIfExternalNavigation": "MidColumnFullScreen",
            "entitySet": "SEPMRA_C_PD_Product",
            "pages": {
              "ObjectPage|to_ProductSalesData": {
                "component": {
                  "name": "sap.suite.ui.generic.template.ObjectPage",
                },
                "defaultLayoutTypeIfExternalNavigation": "MidColumnFullScreen",
                "entitySet": "SEPMRA_C_PD_ProductSalesData",
                "navigationProperty": "to_ProductSalesData",
              },
            },
          },
        },
      },
    },
    "settings": {
      "considerAnalyticalParameters": true,
      "forceGlobalRefresh": false,
      "objectPageHeaderType": "Dynamic",
      "showDraftToggle": false,
    },
  },
  "sap.ui5": {
    "contentDensities": {
      "compact": true,
      "cozy": true,
    },
    "dependencies": {
      "components": {},
      "libs": {
        "sap.suite.ui.generic.template": {
          "lazy": false,
        },
        "sap.ui.core": {
          "lazy": false,
        },
        "sap.ui.generic.app": {
          "lazy": false,
        },
        "se.mi.plm.attachmentservice": {
          "lazy": false,
        },
      },
      "minUI5Version": "1.65.0",
    },
    "extends": {
      "extensions": {},
    },
    "models": {
      "": {
        "dataSource": "mainService",
        "preload": true,
        "settings": {
          "defaultBindingMode": "TwoWay",
          "defaultCountMode": "Inline",
          "metadataUrlParams": {
            "sap-value-list": "none",
          },
          "refreshAfterChange": false,
        },
      },
      "@i18n": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/i18n.properties",
      },
      "i18n": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/i18n.properties",
      },
      "i18n|sap.suite.ui.generic.template.ListReport|SEPMRA_C_PD_Product": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/ListReport/SEPMRA_C_PD_Product/i18n.properties",
      },
      "i18n|sap.suite.ui.generic.template.ObjectPage|SEPMRA_C_PD_Product": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/ObjectPage/SEPMRA_C_PD_Product/i18n.properties",
      },
      "i18n|sap.suite.ui.generic.template.ObjectPage|SEPMRA_C_PD_ProductSalesData": {
        "type": "sap.ui.model.resource.ResourceModel",
        "uri": "i18n/ObjectPage/SEPMRA_C_PD_ProductSalesData/i18n.properties",
      },
    },
    "resources": {
      "css": [],
      "js": [],
    },
  },
}
`;